src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs

```diff
@@ -141,6 +142,30 @@ private Task OnProjectFactoryCompletedAsync()
             }
         }
 
+        private JoinableTask<T> ExecuteWithinLockAsync<T>(Func<Task<T>> task)
+        {
+            return JoinableFactory.RunAsync(async delegate
+            {
+                using (JoinableCollection.Join())
```

Comment:
I agree with Raul that some comments will help other to understand it (and to prevent this in other code too.)  The change looks good to me.